### PR TITLE
KAFKA-12305: Fix Flatten SMT for array types

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-12305)

(Copied from Jira):

The `Flatten` SMT fails for array types. A sophisticated approach that tries to flatten arrays might be desirable in some cases, and may have been punted during the early design phase of the transform, but in the interim, it's probably not worth it to make array data and the SMT mutually exclusive.

A naive approach that preserves arrays as-are and doesn't attempt to flatten them seems fair for now.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by C0urante** [comment]
@gharris1727 @ncliang @tombentley anyone got time to take a look?

**Review by ncliang — APPROVED** [review]
Thanks @C0urante . LGTM

**Review by tombentley — COMMENTED** [review]
Is it worth having a test to cover the recursive flatten case discussed in the JIRA? 

I think we should also elaborate on the behaviour for arrays in the SMT doc, so people know what the SMT is intended to do with arrays (it was never a problem before when the leaves were scalars).

**Comment by C0urante** [comment]
Thanks @tombentley--good call on updating the docs; I've done that. RE a test for the recursive case--I don't think it'll hurt, so I updated the existing tests to include it. Probably not too valuable right now since arrays are just passed through unmodified but it might save someone else that bit of legwork in the future if we decide we want bona fide flattening for array types as well.

**Review by tombentley — APPROVED** [review]
Thanks @C0urante, LGTM. I'll merge it assuming the CI doesn't throw up anything.

**Comment by C0urante** [comment]
@tombentley know it's been a while and we're probably past the boat for 3.0, but wanted to check in and see if there's anything blocking this PR from being merged at the moment?

**Comment by tombentley** [comment]
@C0urante sorry for the delay.

**Pull request merged**
